doc/designs/evp_skey.md

```diff
@@ -0,0 +1,185 @@
+
+```C
+EVP_SKEY * EVP_SKEY_new(void);
+void EVP_SKEY_free(EVP_SKEY *key);
+```
+
+Importing and exporting parameters
+----------------------------------
+
+```C
+int EVP_SKEY_import(EVP_CIPHER *cipher, EVP_SKEY **pskey, OSSL_PARAM params[]);
+int EVP_SKEY_export(const EVP_SKEY *key, OSSL_PARAM **params,
+                    OSSL_CALLBACK *export_cb, void *export_cbarg);
+```
+
+A parameter `raw_bytes` can be used if we try to get a key in classical raw
+bytes implementation.
+
+To find out the parameters supported by a particular EVP_CIPHER object,
+we need a function
+
+```C
+const OSSL_PARAM *EVP_SKEY_fromdata_settable(EVP_CIPHER *cipher);
+```
+
+similar to `EVP_PKEY_fromdata_settable`
+
+API to derive a symmetric key
+-----------------------------
+
+The derived key can be algorithm-specific or algorithm-agnostic. To specify the
+algorithm binding, the params argument can be used.
+
+```C
+int EVP_PKEY_derive_SKEY(EVP_PKEY_CTX *ctx, EVP_SKEY *pskey,
+                         OSSL_PARAM params[]);
+int EVP_KDF_derive_SKEY(EVP_PKEY_CTX *ctx, EVP_SKEY *pskey,
+                        OSSL_PARAM params[]);
+```
+
+similar to `EVP_PKEY_derive/EVP_KDF_derive`
+
+Duplicating the EVP_SKEY object
+-------------------------------
+
+**Open question:** do we need it?
+
+```C
+EVP_SKEY * EVP_SKEY_dup(const EVP_SKEY * skey);
+```
+
+Using in cipher operations
+--------------------------
+
+We already have the function
+
+```C
+EVP_CipherInit_ex2(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
+                   const unsigned char *key, const unsigned char *iv,
+                   int enc, const OSSL_PARAM params[]);
+```
+
+To set the skey object to the context, we invoke `EVP_CipherInit_ex2` with NULL
+key ptr and invoke `EVP_CIPHER_CTX_set1_EVP_SKEY`.
+
+An alternate approach will be providing a function
+
+```C
+EVP_CipherInit_skey(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
+                   const EVP_SKEY *key, const unsigned char *iv,
+                   int enc, const OSSL_PARAM ctx_params[]);
+```
+
+altogether with `EVP_EncryptInit_skey/EVP_DecryptInit_skey` wrappers.
+
+Accessing the current key object
+--------------------------------
+
+To get/set the currently set key as a EVP_SKEY object, we introduce the API
+
+```C
+EVP_SKEY * EVP_CIPHER_CTX_get1_EVP_SKEY(const EVP_CIPHER_CTX *ctx, int format);
+EVP_SKEY * EVP_MAC_CTX_get1_EVP_SKEY(const EVP_MAC_CTX *ctx, EVP_SKEY *skey,
+                                     int format);
+EVP_SKEY * EVP_KDF_CTX_get1_EVP_SKEY(const EVP_KDF_CTX *ctx, EVP_SKEY *skey,
+                                     int format);
+int EVP_CIPHER_CTX_set1_EVP_SKEY(const EVP_CIPHER_CTX *ctx, EVP_SKEY *skey);
+int EVP_MAC_CTX_set1_EVP_SKEY(const EVP_MAC_CTX *ctx, EVP_SKEY *skey);
+int EVP_KDF_CTX_set1_EVP_SKEY(const EVP_KDF_CTX *ctx, EVP_SKEY *skey);
+```
+
+The `format` argument of getters specifies whether we try to obtain raw bytes
+implementation or provider-backed opaque handle fits well for our purposes.
```

Comment:
Why is a `format` necessary?

Reply:
We may want to avoid conversion from/to opaque representation if we use the same cipher object.

Reply:
Same as...  what?  I'd suggest that when getting the `EVP_SKEY` from a context associated with a provider, it should get the form that's convenient for that provider.  Conversion would happen *later*, no? ... when trying to use the resulting EVP_SKEY with *another* provider

Reply:
Fair point. Removing.